### ISSUE TITLE
[BUG] Jax device interface leaks tracers when defining devices inside of jitted-function

### ISSUE THREAD

**Issue by PhilipVinc** [PennyLaneAI/pennylane#3205]
### Expected behavior

Expected: The following code to run 😁. It uses jax leak checker to verify that some code is correct, and the checker says that something is bing leaked.

This is a MWE reduced by me and @StefanoBarison that arose from some more complicated code where we were trying to use  finite number of shots. 

While this leak is harmless in this particular MWE, I suspect that it creates problems when using shots and computing gradients.

### Additional information

If you inspect the stack trace, you'll notice that it says that one of the tracers leaked was created in, which is the state stored inside the device. As I am not returning the device itself, may it be that you are caching devices constructed somewhere?

```
This Tracer was created on line .../pennylane/devices/default_qubit.py:653 (_create_basis_state)
```

Also, apparently I leak two more tracers from the line calling `_measure_operator` but that probably is because the leak is happening in penny lane's tape and this is confusing jax stack trace.
```
Traced<ShapedArray(float32[])>with<DynamicJaxprTrace(level=0/1)>
This Tracer was created on line .../shots_MWE2.py:18 (_measure_operator)
```

Tested both on current latest stable release and GitHub master branch. The error thrown is the same.
```python
>>> jax.__version__
'0.3.23'
```

### Source code

```python
from typing import Tuple, Optional
from functools import partial

import jax
import jax.numpy as jnp
import numpy as np

import pennylane as qml
from flax import struct

phys_qubits = 2
pars_q      = np.random.rand(2)

def minimal_circ(params):
    dev = qml.device("default.qubit.jax", wires=tuple(range(phys_qubits)), shots=None)
    @qml.qnode(dev, interface="jax",diff_method="parameter-shift")
    def _measure_operator():
        qml.RY(params[0],wires=0)
        qml.RY(params[1],wires=1)
        op = qml.Hamiltonian([1.0],[qml.PauliZ(0) @ qml.PauliZ(1)])
        return qml.expval(op)
    res = _measure_operator()
    print("res is: ", res, type(res))
    return res

print("================== don't check leaks: works ========================")
out = minimal_circ(pars_q)
print(f"{out=}")

print("================== check leaks: cry ========================")
with jax.checking_leaks():
    out = jax.jit(minimal_circ)(pars_q)
    print(f"{out=}")
```

### Tracebacks

```shell
➜ python shots_MWE2.py
================== don't check leaks: works ========================
res is:  0.67008924 <class 'jaxlib.xla_extension.DeviceArray'>
out=DeviceArray(0.67008924, dtype=float32)
================== check leaks: cry ========================
res is:  Traced<ShapedArray(float32[])>with<DynamicJaxprTrace(level=0/1)> <class 'jax.interpreters.partial_eval.DynamicJaxprTracer'>
Traceback (most recent call last):
  File "/Users/filippovicentini/Dropbox/Ricerca/Cartelle_Condivise/qng_and_friends/embedding/shots_MWE2.py", line 32, in <module>
    out = jax.jit(minimal_circ)(pars_q)
  File "/Users/filippovicentini/Documents/pythonenvs/qng-friends/python-3.10.6/lib/python3.10/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/Users/filippovicentini/Documents/pythonenvs/qng-friends/python-3.10.6/lib/python3.10/site-packages/jax/_src/api.py", line 606, in cache_miss
    out_flat = xla.xla_call(
  File "/Users/filippovicentini/Documents/pythonenvs/qng-friends/python-3.10.6/lib/python3.10/site-packages/jax/core.py", line 1939, in bind
    return call_bind(self, fun, *args, **params)
  File "/Users/filippovicentini/Documents/pythonenvs/qng-friends/python-3.10.6/lib/python3.10/site-packages/jax/core.py", line 1955, in call_bind
    outs = top_trace.process_call(primitive, fun_, tracers, params)
  File "/Users/filippovicentini/Documents/pythonenvs/qng-friends/python-3.10.6/lib/python3.10/site-packages/jax/core.py", line 701, in process_call
    return primitive.impl(f, *tracers, **params)
  File "/Users/filippovicentini/Documents/pythonenvs/qng-friends/python-3.10.6/lib/python3.10/site-packages/jax/_src/dispatch.py", line 234, in _xla_call_impl
    compiled_fun = xla_callable(fun, device, backend, name, donated_invars,
  File "/Users/filippovicentini/Documents/pythonenvs/qng-friends/python-3.10.6/lib/python3.10/site-packages/jax/linear_util.py", line 309, in memoized_fun
    ans = call(fun, *args)
  File "/Users/filippovicentini/Documents/pythonenvs/qng-friends/python-3.10.6/lib/python3.10/site-packages/jax/_src/dispatch.py", line 342, in _xla_callable_uncached
    return lower_xla_callable(fun, device, backend, name, donated_invars, False,
  File "/Users/filippovicentini/Documents/pythonenvs/qng-friends/python-3.10.6/lib/python3.10/site-packages/jax/_src/profiler.py", line 313, in wrapper
    return func(*args, **kwargs)
  File "/Users/filippovicentini/Documents/pythonenvs/qng-friends/python-3.10.6/lib/python3.10/site-packages/jax/_src/dispatch.py", line 428, in lower_xla_callable
    jaxpr, out_type, consts = pe.trace_to_jaxpr_final2(
  File "/Users/filippovicentini/Documents/pythonenvs/qng-friends/python-3.10.6/lib/python3.10/site-packages/jax/_src/profiler.py", line 313, in wrapper
    return func(*args, **kwargs)
  File "/Users/filippovicentini/Documents/pythonenvs/qng-friends/python-3.10.6/lib/python3.10/site-packages/jax/interpreters/partial_eval.py", line 2079, in trace_to_jaxpr_final2
    with core.new_sublevel():
  File "/Users/filippovicentini/.pyenv/versions/3.10.6/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/contextlib.py", line 142, in __exit__
    next(self.gen)
  File "/Users/filippovicentini/Documents/pythonenvs/qng-friends/python-3.10.6/lib/python3.10/site-packages/jax/core.py", line 1032, in new_sublevel
    raise leaked_tracer_error("sublevel", t(), leaked_tracers)
jax._src.traceback_util.UnfilteredStackTrace: Exception: Leaked sublevel 1. Leaked tracer(s):

Traced<ShapedArray(complex64[2,2])>with<DynamicJaxprTrace(level=0/1)>
This Tracer was created on line /Users/filippovicentini/Documents/pythonenvs/qng-friends/python-3.10.6/lib/python3.10/site-packages/pennylane/devices/default_qubit.py:653 (_create_basis_state)

Traced<ShapedArray(float32[])>with<DynamicJaxprTrace(level=0/1)>
This Tracer was created on line /Users/filippovicentini/Dropbox/Ricerca/Cartelle_Condivise/qng_and_friends/embedding/shots_MWE2.py:18 (_measure_operator)

Traced<ShapedArray(float32[])>with<DynamicJaxprTrace(level=0/1)>
This Tracer was created on line /Users/filippovicentini/Dropbox/Ricerca/Cartelle_Condivise/qng_and_friends/embedding/shots_MWE2.py:18 (_measure_operator)

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/filippovicentini/Dropbox/Ricerca/Cartelle_Condivise/qng_and_friends/embedding/shots_MWE2.py", line 32, in <module>
    out = jax.jit(minimal_circ)(pars_q)
  File "/Users/filippovicentini/.pyenv/versions/3.10.6/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/contextlib.py", line 142, in __exit__
    next(self.gen)
Exception: Leaked sublevel 1. Leaked tracer(s):

Traced<ShapedArray(complex64[2,2])>with<DynamicJaxprTrace(level=0/1)>
This Tracer was created on line /Users/filippovicentini/Documents/pythonenvs/qng-friends/python-3.10.6/lib/python3.10/site-packages/pennylane/devices/default_qubit.py:653 (_create_basis_state)

Traced<ShapedArray(float32[])>with<DynamicJaxprTrace(level=0/1)>
This Tracer was created on line /Users/filippovicentini/Dropbox/Ricerca/Cartelle_Condivise/qng_and_friends/embedding/shots_MWE2.py:18 (_measure_operator)

Traced<ShapedArray(float32[])>with<DynamicJaxprTrace(level=0/1)>
This Tracer was created on line /Users/filippovicentini/Dropbox/Ricerca/Cartelle_Condivise/qng_and_friends/embedding/shots_MWE2.py:18 (_measure_operator)
```

### System information

```shell
>>> import pennylane as qml; qml.about()
Name: PennyLane
Version: 0.27.0.dev0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author:
Author-email:
License: Apache License 2.0
Location: /Users/filippovicentini/Documents/pythonenvs/qng-friends/python-3.10.6/lib/python3.10/site-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, retworkx, scipy, semantic-version, toml
Required-by: PennyLane-Lightning

Platform info:           macOS-12.6-x86_64-i386-64bit
Python version:          3.10.6
Numpy version:           1.23.4
Scipy version:           1.9.3
Installed devices:
- default.gaussian (PennyLane-0.27.0.dev0)
- default.mixed (PennyLane-0.27.0.dev0)
- default.qubit (PennyLane-0.27.0.dev0)
- default.qubit.autograd (PennyLane-0.27.0.dev0)
- default.qubit.jax (PennyLane-0.27.0.dev0)
- default.qubit.tf (PennyLane-0.27.0.dev0)
- default.qubit.torch (PennyLane-0.27.0.dev0)
- default.qutrit (PennyLane-0.27.0.dev0)
- null.qubit (PennyLane-0.27.0.dev0)
- lightning.qubit (PennyLane-Lightning-0.26.1)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by antalszava** [comment]
Hi @PhilipVinc, thank you so much for reporting this! :+1:  We'll be taking a closer look.

**Comment by antalszava** [comment]
@PhilipVinc thanks again for the report!

We've investigated this, but will likely have to keep having this on our radar. PennyLane's implementation originally didn't consider a purely functional approach and for that reason, likely leaking tracers happens on several occasions.

In general, we're keen to hear more about the shortcomings of leaking tracers and would like to do a memory profiling while running an optimization workflow in PennyLane with JAX computation leaking tracers to better understand the problem. Any insights are appreciated!

Specifically, the above linked example seems to reduce to
```py
pars_q      = jnp.ones(2)
dev = qml.device("default.qubit.jax", wires=2, shots=None)
def minimal_circ(params, dev):
    qs = qml.tape.QuantumTape([qml.RY(params[0],wires=0), qml.RY(params[1],wires=1)],[qml.expval(qml.PauliZ(0))])
    return dev.execute(qs)

with jax.checking_leaks():
    jax.jit(minimal_circ, static_argnums=[1])(pars_q, dev)
```
```
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "leak.py", line 19, in <module>
    jax.jit(minimal_circ, static_argnums=[1])(pars_q, dev)
  File "/home/antal/anaconda3/lib/python3.8/contextlib.py", line 120, in __exit__
    next(self.gen)
Exception: Leaked sublevel 1. Leaked tracer(s):

Traced<ShapedArray(complex64[2,2])>with<DynamicJaxprTrace(level=0/1)>
This Tracer was created on line /home/antal/anaconda3/lib/python3.8/contextlib.py:75 (inner)
```
Frankly, the output error message is not at all descriptive, so it's challenging to track down the exact cause. Any tips on doing so are appreciated.

Posting another example that has a slightly more descriptive error message as it outputs info about the leaked tracer:
```py
pars_q      = jnp.ones(2)
dev = qml.device("default.qubit.jax", wires=2, shots=None)
def minimal_circ(params, dev):
    qs = qml.tape.QuantumTape([qml.RY(params[0],wires=0), qml.RY(params[1],wires=1)],[qml.expval(qml.PauliZ(0))])
    return dev.execute(qs)

with jax.checking_leaks():
    jax.jacrev(minimal_circ)(pars_q, dev)
```
```
  File "/home/antal/anaconda3/lib/python3.8/site-packages/jax/interpreters/partial_eval.py", line 768, in trace_to_jaxpr_nounits
    jaxpr, (out_pvals, consts, env) = fun.call_wrapped(pvals)
  File "/home/antal/anaconda3/lib/python3.8/site-packages/jax/linear_util.py", line 183, in call_wrapped
    ans = gen.send(ans)
  File "/home/antal/anaconda3/lib/python3.8/site-packages/jax/interpreters/ad.py", line 73, in jvpfun
    del main
  File "/home/antal/anaconda3/lib/python3.8/contextlib.py", line 120, in __exit__
    next(self.gen)
  File "/home/antal/anaconda3/lib/python3.8/site-packages/jax/core.py", line 928, in new_main
    if leaked_tracers: raise leaked_tracer_error("trace", t(), leaked_tracers)
Exception: Leaked trace MainTrace(2,JVPTrace). Leaked tracer(s):

Traced<ConcreteArray([[0.77015114+0.j 0.4207355 +0.j]
 [0.4207355 +0.j 0.22984886+0.j]], dtype=complex64)>with<JVPTrace(level=2/0)> with
  primal = DeviceArray([[0.77015114+0.j, 0.4207355 +0.j],
             [0.4207355 +0.j, 0.22984886+0.j]], dtype=complex64)
  tangent = Traced<ShapedArray(complex64[2,2])>with<JaxprTrace(level=1/0)> with
    pval = (ShapedArray(complex64[2,2]), None)
    recipe = JaxprEqnRecipe(eqn_id=<object object at 0x7fe4fc70ef30>, in_tracers=(Traced<ConcreteArray([[ 0.87758255+0.j -0.47942555-0.j]
 [ 0.47942555+0.j  0.87758255+0.j]], dtype=complex64):JaxprTrace(level=1/0)>, Traced<ConcreteArray([[0.87758255+0.j 0.        +0.j]
 [0.47942555+0.j 0.        +0.j]], dtype=complex64):JaxprTrace(level=1/0)>, Traced<ShapedArray(complex64[2,2]):JaxprTrace(level=1/0)>, Traced<ShapedArray(complex64[2,2]):JaxprTrace(level=1/0)>), out_tracer_refs=[<weakref at 0x7fe4f8249e50; to 'JaxprTracer' at 0x7fe4f8249f40>], out_avals=[ShapedArray(complex64[2,2])], primitive=xla_call, params={'device': None, 'backend': None, 'name': '_einsum', 'donated_invars': (False, False, False, False), 'inline': False, 'keep_unused': False, 'call_jaxpr': { lambda ; a:c64[2,2] b:c64[2,2] c:c64[2,2] d:c64[2,2]. let
    e:c64[2,2] = dot_general[
      dimension_numbers=(((1,), (1,)), ((), ()))
      precision=None
      preferred_element_type=None
    ] d a
    f:c64[2,2] = dot_general[
      dimension_numbers=(((1,), (1,)), ((), ()))
      precision=None
      preferred_element_type=None
    ] b c
    g:c64[2,2] = add_any e f
  in (g,) }}, effects=set(), source_info=SourceInfo(traceback=<jaxlib.xla_extension.Traceback object at 0x7fe4fcac0d30>, name_stack=NameStack(stack=(Transform(name='jvp'), Scope(name='...cb,...ab->...ac')))))
```

**Comment by PhilipVinc** [comment]
@antalszava ok, one issue I see with **your MWE** is that `JaxDevice` has a field `_prngkey` which is in principle a jax array, but jax does not know about it (this might be a source of leaks).

In principle, in jax, you have to differentiate between static and dynamical data. Dynamical data are arrays (like a prng key) which should not retrigger compilation/tracing when it changes (eg: you don't recompile every time you feed a matmul a different matrix). Static data is anything that, when it changes, it should retrigger compilation/tracing.

Once you call a `jax.***` function with an object, jax automatically attempts to unpack it into `static` and `dynamic` parts. Some parts of the API are loud about errors in this (eg: jit) some parts are less loud. All objects you wish to pass to a jax api should be splittable as defined in [here](https://jax.readthedocs.io/en/latest/pytrees.html), but your devices are not, so you should **never** pass them at a jax API boundary.

In particular, I think your devices could be 99% ok as static data, except for the prngkey field...

**Comment by PhilipVinc** [comment]
You are capturing the `device`, which contains the tracer `device._prng_key` in [`jax_jit.py:_execute](https://github.com/PennyLaneAI/pennylane/blob/f5d1e68e49b103ea95fc5c8adf39234eb62fa4dc/pennylane/interfaces/jax_jit.py#L153).

This leaks because jax keeps a list of those functions/partials so this object survives after the end of tracing.

This is also what causes a crash in my gist, as in the [callback](https://github.com/PennyLaneAI/pennylane/blob/f5d1e68e49b103ea95fc5c8adf39234eb62fa4dc/pennylane/interfaces/jax_jit.py#L183)  I suppose you are calling some code that contains tracers.

You should be passing the prng key as an argument of the callback like `params` is. ( I admit my suggestion is vague... but It's the first time I'm diving so deep into penny lane source)

**Comment by PhilipVinc** [comment]
Closing in favour of #3218 . This is a different issue but this one is harmless. I will reopen with a better investigation.

**Comment by antalszava** [comment]
One thing we investigated just now with Sam: looks like the issue might be with contraction each PennyLane performs (be it einsum or tensordot).

`default.qubit` would for example do something analogous to `state = jnp.tensordot( mat, state, axes)` where `mat` is a Tracer, but `state` is a `DeviceArray`. In itself, this should not cause an issue, yet, debugging the related parts of `contextlib` starts to point into this direction.
```py
Traced<ShapedArray(complex64[2,2])>with<DynamicJaxprTrace(level=0/1)> [[1.+0.j 0.+0.j]
 [0.+0.j 0.+0.j]] [(-1,), (0,)] <class 'jaxlib.xla_extension.DeviceArray'>
```

**Comment by PhilipVinc** [comment]
@antalszava if you pip install this version https://github.com/google/jax/pull/13022/files of jax
```
pip install git+https://github.com/mattjj/jax@leak-checker-improvements
```

you will get much more understandable output like

```python

Traced<ShapedArray(complex64[2,2])>with<DynamicJaxprTrace(level=0/1)>
This DynamicJaxprTracer was created on line /Users/filippovicentini/Dropbox/Ricerca/Codes/Python/pennylane/pennylane/devices/default_qubit.py:653 (_create_basis_state)
<DynamicJaxprTracer 5431981488> is alive because it is referred to by <DefaultQubitJax 5432087488> in vars '_state', '_pre_rotated_state'
<DefaultQubitJax 5432087488> is alive because it is referred to by <tuple 5431306944>[1]
<tuple 5431306944> is alive because it is referred to by <_lru_cache_wrapper 5400196928>
<_lru_cache_wrapper 5400196928> is alive because it is referred to by <dict 5400182400>['shape']
<dict 5400182400> is alive because it is referred to by <type 140602542618192>
<type 140602542618192> is alive because it is referred to by pennylane.measurements.MeasurementProcess

Traced<ShapedArray(float32[])>with<DynamicJaxprTrace(level=0/1)>
This DynamicJaxprTracer was created on line /Users/filippovicentini/Dropbox/Ricerca/Codes/Python/pennylane/pp2.py:14 (_measure_operator)
<DynamicJaxprTracer 5432135408> is alive because it is referred to by <list 5432157248>[0]
<list 5432157248> is alive because it is referred to by <RY 5432087824>.data
<RY 5432087824> is alive because it is referred to by <OrderedDict 5430958272> at keys
<OrderedDict 5430958272> is alive because it is referred to by <QuantumTape 5432087728>._queue
<QuantumTape 5432087728> is alive because it is referred to by <list 5432155264>[0]
<list 5432155264> is alive because it is referred to by <function 5432110688> closed-over variable tapes
<function 5432110688> is alive because it is referred to by <WrappedFun 5432140688>
<WrappedFun 5432140688> is alive because it is referred to by <function 5432111408> closed-over variable bwd
<function 5432111408> is alive because it is referred to by <WrappedFun 5432302608>
<WrappedFun 5432302608> is alive because it is referred to by <dict 5431394176>['bwd']
<dict 5431394176> is alive because it is referred to by <JaxprEqn 5432281984>[3]
<JaxprEqn 5432281984> is alive because it is referred to by <list 5432259520>[5]
<list 5432259520> is alive because it is referred to by <Jaxpr 5432095648>.eqns

Traced<ShapedArray(float32[])>with<DynamicJaxprTrace(level=0/1)>
This DynamicJaxprTracer was created on line /Users/filippovicentini/Dropbox/Ricerca/Codes/Python/pennylane/pp2.py:14 (_measure_operator)
<DynamicJaxprTracer 5432134528> is alive because it is referred to by <list 5432159040>[0]
<list 5432159040> is alive because it is referred to by <RY 5432087680>.data
<RY 5432087680> is alive because it is referred to by <OrderedDict 5430958272> at keys
<OrderedDict 5430958272> is alive because it is referred to by <QuantumTape 5432087728>._queue
<QuantumTape 5432087728> is alive because it is referred to by <list 5432155264>[0]
<list 5432155264> is alive because it is referred to by <function 5432110688> closed-over variable tapes
<function 5432110688> is alive because it is referred to by <WrappedFun 5432140688>
<WrappedFun 5432140688> is alive because it is referred to by <function 5432111408> closed-over variable bwd
<function 5432111408> is alive because it is referred to by <WrappedFun 5432302608>
<WrappedFun 5432302608> is alive because it is referred to by <dict 5431394176>['bwd']
<dict 5431394176> is alive because it is referred to by <JaxprEqn 5432281984>[3]
<JaxprEqn 5432281984> is alive because it is referred to by <list 5432259520>[5]
<list 5432259520> is alive because it is referred to by <Jaxpr 5432095648>.eqns

```
